# Step 1: Define the Goal

To transform the raw_facebook_data.csv based on the EDA.

# Step 2: Load the Data

## Load the data

Import packages

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

Read the original CSV file and define the data frame as df

In [2]:
df = pd.read_csv('../data/raw_facebook_data.csv', sep=";")
df.head()

,Page total likes,Type,Category,Post Month,Post Weekday,Post Hour,Paid,Lifetime Post Total Reach,Lifetime Post Total Impressions,Lifetime Engaged Users,Lifetime Post Consumers,Lifetime Post Consumptions,Lifetime Post Impressions by people who have liked your Page,Lifetime Post reach by people who like your Page,Lifetime People who have liked your Page and engaged with your post,comment,like,share,Total Interactions
0,139441,Photo,2,12,4,3,0.0,2752,5091,178,109,159,3078,1640,119,4,79.0,17.0,100
1,139441,Status,2,12,3,10,0.0,10460,19057,1457,1361,1674,11710,6112,1108,5,130.0,29.0,164
2,139441,Photo,3,12,3,3,0.0,2413,4373,177,113,154,2812,1503,132,0,66.0,14.0,80
3,139441,Photo,2,12,2,10,1.0,50128,87991,2211,790,1119,61027,32048,1386,58,1572.0,147.0,1777
4,139441,Photo,2,12,2,3,0.0,7244,13594,671,410,580,6228,3200,396,19,325.0,49.0,393


# Step 2: Transform the data

Tasks:
- Handling missing values
- Fixing incorrect data types
- Drop irrelvant columns
- Log transformation


## Handling missing values

In [3]:
df = df.dropna(subset=['Paid'])
df[df['Paid'].isna()].shape

(0, 19)

Index 499 has been dropped

In [4]:
df.shape

(499, 19)

In [5]:
df['like'] = df['like'].fillna(df['like'].mean(numeric_only=True))
df[df['like'].isna()].shape

(0, 19)

In [6]:
df['share'] = df['share'].fillna(df['share'].mean(numeric_only=True))
df[df['share'].isna()].shape

(0, 19)

Index 111, 120, 124, and 164 have been filled with appropriate replacements

In [7]:
df.shape

(499, 19)

No other rows have been dropped during the fillna process

## Fix incorrect data types 

The only column that needs to be converted from a categorical one to numerical is the "Category" column. 

In [8]:
df["Type"].value_counts()

Type
Photo     425
Status     45
Link       22
Video       7
Name: count, dtype: int64

In [9]:
type_dummies = pd.get_dummies(df['Type'], dtype=int, drop_first=True)
type_dummies
df.drop('Type', axis=1, inplace=True) 
df = df.join(type_dummies)
df.shape

(499, 21)

In [10]:
df.head()

,Page total likes,Category,Post Month,Post Weekday,Post Hour,Paid,Lifetime Post Total Reach,Lifetime Post Total Impressions,Lifetime Engaged Users,Lifetime Post Consumers,Lifetime Post Consumptions,Lifetime Post Impressions by people who have liked your Page,Lifetime Post reach by people who like your Page,Lifetime People who have liked your Page and engaged with your post,comment,like,share,Total Interactions,Photo,Status,Video
0,139441,2,12,4,3,0.0,2752,5091,178,109,159,3078,1640,119,4,79.0,17.0,100,1,0,0
1,139441,2,12,3,10,0.0,10460,19057,1457,1361,1674,11710,6112,1108,5,130.0,29.0,164,0,1,0
2,139441,3,12,3,3,0.0,2413,4373,177,113,154,2812,1503,132,0,66.0,14.0,80,1,0,0
3,139441,2,12,2,10,1.0,50128,87991,2211,790,1119,61027,32048,1386,58,1572.0,147.0,1777,1,0,0
4,139441,2,12,2,3,0.0,7244,13594,671,410,580,6228,3200,396,19,325.0,49.0,393,1,0,0


Using get_dummies method, we converted the values from the "Type" column into 3 new columns with binary values: "Photo," "Status," and "Video." Note that we dropped one of the dummy columns ("Link") with dropfirst=True to avoid the Dummy Variable Trap caused by the multicolinary between the related variables.

# Drop irrelvant columns

In [11]:
df = df.drop(['comment', 'like', 'share'], axis=1)
df.head()

,Page total likes,Category,Post Month,Post Weekday,Post Hour,Paid,Lifetime Post Total Reach,Lifetime Post Total Impressions,Lifetime Engaged Users,Lifetime Post Consumers,Lifetime Post Consumptions,Lifetime Post Impressions by people who have liked your Page,Lifetime Post reach by people who like your Page,Lifetime People who have liked your Page and engaged with your post,Total Interactions,Photo,Status,Video
0,139441,2,12,4,3,0.0,2752,5091,178,109,159,3078,1640,119,100,1,0,0
1,139441,2,12,3,10,0.0,10460,19057,1457,1361,1674,11710,6112,1108,164,0,1,0
2,139441,3,12,3,3,0.0,2413,4373,177,113,154,2812,1503,132,80,1,0,0
3,139441,2,12,2,10,1.0,50128,87991,2211,790,1119,61027,32048,1386,1777,1,0,0
4,139441,2,12,2,3,0.0,7244,13594,671,410,580,6228,3200,396,393,1,0,0


## Log transformation

In [12]:
exclude_cols = ["Paid", "Photo", "Status", "Video", "Total Interactions"]
numeric_cols = df.select_dtypes(include=[np.number]).columns.difference(exclude_cols)
df_log = df.copy()
df_log[numeric_cols] = np.log1p(df[numeric_cols])  # Transform only continuous numerics

df_log.head()

,Page total likes,Category,Post Month,Post Weekday,Post Hour,Paid,Lifetime Post Total Reach,Lifetime Post Total Impressions,Lifetime Engaged Users,Lifetime Post Consumers,Lifetime Post Consumptions,Lifetime Post Impressions by people who have liked your Page,Lifetime Post reach by people who like your Page,Lifetime People who have liked your Page and engaged with your post,Total Interactions,Photo,Status,Video
0,11.845404,1.098612,2.564949,1.609438,1.386294,0.0,7.920447,8.535426,5.187386,4.700480,5.075174,8.032360,7.403061,4.787492,100,1,0,0
1,11.845404,1.098612,2.564949,1.386294,2.397895,0.0,9.255409,9.855242,7.284821,7.216709,7.423568,9.368284,8.718173,7.011214,164,0,1,0
2,11.845404,1.386294,2.564949,1.386294,1.386294,0.0,7.789040,8.383433,5.181784,4.736198,5.043425,7.942007,7.315884,4.890349,80,1,0,0
3,11.845404,1.098612,2.564949,1.098612,2.397895,1.0,10.822355,11.385001,7.701652,6.673298,7.021084,11.019088,10.375021,7.234898,1777,1,0,0
4,11.845404,1.098612,2.564949,1.098612,1.386294,0.0,8.888067,9.517457,6.510258,6.018593,6.364751,8.736971,8.071219,5.983936,393,1,0,0


# Step 3: Final Check

In [13]:
print(df.shape)
print(df.isnull().sum())
df.dtypes

(499, 18)
Page total likes                                                       0
Category                                                               0
Post Month                                                             0
Post Weekday                                                           0
Post Hour                                                              0
Paid                                                                   0
Lifetime Post Total Reach                                              0
Lifetime Post Total Impressions                                        0
Lifetime Engaged Users                                                 0
Lifetime Post Consumers                                                0
Lifetime Post Consumptions                                             0
Lifetime Post Impressions by people who have liked your Page           0
Lifetime Post reach by people who like your Page                       0
Lifetime People who have liked your Page 

Page total likes                                                         int64
Category                                                                 int64
Post Month                                                               int64
Post Weekday                                                             int64
Post Hour                                                                int64
Paid                                                                   float64
Lifetime Post Total Reach                                                int64
Lifetime Post Total Impressions                                          int64
Lifetime Engaged Users                                                   int64
Lifetime Post Consumers                                                  int64
Lifetime Post Consumptions                                               int64
Lifetime Post Impressions by people who have liked your Page             int64
Lifetime Post reach by people who like your Page    

# Step 4: Export transformed data

In [16]:
df_log.to_csv('../data/log_cleaned_facebook_data.csv', index=False)

In [17]:
df.to_csv('../data/original_cleaned_facebook_data.csv', index=False)

Export both clean versions of the log transformed and original data to compare model performance.